File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Sum Root to Leaf Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int ans = 0;
    void sumNumber(TreeNode *root, int sum)
    {
        if (root == nullptr)
        {
            return;
        }
        if (root->left == nullptr && root->right == nullptr)
        {
            ans += (sum * 10 + root->val);
            return;
        }
        if (root->right)
            sumNumber(root->right, root->val + sum * 10);
        if (root->left)
            sumNumber(root->left, root->val + sum * 10);
    }
    int sumNumbers(TreeNode *root)
    {
        ans = 0;
        sumNumber(root, 0);
        return ans;
    }
};

File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Sum Root to Leaf Numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        self.ans = 0
        self.sum(root, 0)
        return self.ans

    def sum(self, root: TreeNode, total: int):
        if root is None:
            return
        if root.left is None and root.right is None:
            self.ans += (total*10 + root.val)
            return
        if root.left:
            self.sum(root.left, total*10 + root.val)
        if root.right:
            self.sum(root.right, total*10 + root.val)
